Classifier dark theme

srallen opened the pull request:
Staging branch URL: https://classifier-dark-theme.pfe-preview.zooniverse.org/

This is the other half for #4470 and is ready for an internal technical and design review. If this is ok, this will merge into #4470 and then that branch will have a final review with the beta test listserv and rebase before merging into master.

I mostly just used Kitteh Zoo and the dev classifier to test with:
https://classifier-dark-theme.pfe-preview.zooniverse.org/projects/brooke/i-fancy-cats/classify
https://classifier-dark-theme.pfe-preview.zooniverse.org/dev/classifier

This does not include:
- Style updates for the drawing subtask as seen in Kitteh Zoo's drawing task:
<img width="1145" alt="screen shot 2018-04-03 at 3 05 53 pm" src="https://user-images.githubusercontent.com/5016731/38273144-d88993ca-3750-11e8-8a30-99a1bc45ede4.png">
- Finished light theme for the survey task (the choice view needs design input) or any work on the dark theme version for the survey task

These two things I think will need a bit of time, so I think best tackled separately again. 

@beckyrother One thing that needs addressing is that the dark theme hover/focus state for the buttons in the task has a thicker boarder, 3px vs 1px, and this will cause a bouncing type effect when the button state changes. Typically this is solved by making sure the border stays the same thickness between states. Do you still want the 3px border on hover/focus in the dark theme?

This PR includes a lot of refactoring of existing components or elements that are now standalone components now using styled-components as to take advantage of its theming plugin. All of the new components should have tests.

I don't think this needs to be reviewed by everyone I requested, just from whoever has time this week hopefully!

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

beckyrother commented:
This looks great! 

On the focus border, yes, both should be 3px. Can the border also be #5C5C5C? It's quite contrast-y now to the point of being a bit distracting. 

Other than that, everything looks awesome.

eatyourgreens commented:
One small thing: on my test project, I can click the Back button on the first task, even though I can't go back. https://classifier-dark-theme.pfe-preview.zooniverse.org/projects/eatyourgreens/-project-testing-ground/classify

eatyourgreens commented on app/classifier/tasks/components/TaskInputField/components/TaskInputLabel/TaskInputLabel.jsx:
```diff
@@ -0,0 +1,71 @@
+import styled from 'styled-components';
+
+function howShouldTheLabelBeAligned(label, labelIcon) {
+  if ((label && label.includes('<img')) || (label && labelIcon)) {
```
Is the label HTML or Markdown? The label strings in task translations are Markdown, and images are in square brackets (I think.)

srallen replied:
Good point. I'll check on that.

eatyourgreens replied:
Maybe the Markdown was parsed further up the chain, in which case ignore my comment.

srallen commented:
I just rebased this to resolve the conflicts. There's a styling issue now with the dev classifier due to the recent refactoring to handle annotations prop. I also still need to address the style comments.

srallen commented:
@beckyrother that'll be great if you have the time. I can muddle through otherwise 👍

beckyrother commented:
Et voici: https://invis.io/F3F2T5K2Q#/291878117_000_Answers_With_An_Image (also updating the google doc with this link)

Dark version: https://invis.io/F3F2T5K2Q#/291878845_08_Dark_with_Image

srallen commented:
@eatyourgreens I've checked the performance on the event handlers on the inputs. It looks like the total time for keydown, blur, focus, and click is about 50ms and the dev tools are warning about that time. I haven't seen it skip showing the UI update though. Could you let me know what the performance dev tools is showing you? Is it taking much more time for you?

eatyourgreens commented:
@srallen it's really slow for me on the initial question in the dev classifier. If I hold the down arrow down for a few seconds, the task selection becomes really jerky, skipping 2 or 3 choices before changing the current selection. I'll have a look at the timing.

eatyourgreens commented:
@srallen according to the dev tools, I'm seeing ~150ms for the keydown handler.

srallen commented:
I wonder why it's taking 3x as long to fire for you. 🤔

beckyrother commented:
**Design comments, all:**
- Images in the answers could use ~5px of padding on all sides
- Need to add some feedback when an answer is selected and still hovered-over. An "active" state? on [this page](https://invis.io/F3F2T5K2Q#/270495246_01-1_Question_Selected), the dark teal outline is not in the correct place, but illustrates what I'm talking about – when you click on an answer, the dark outline should appear to indicate that the answer has been selected. Please let me know if that's unclear, it's difficult to describe.

**Dark theme:**
- Can the hover state for the answers have the same style as the hover state for the Back button? filled in with dark teal
- Border should be consistent width through all states

srallen commented:
@eatyourgreens does the keydown event show a drop down arrow where you can check into the other events fired? Is the 150ms total an even sum or is there a particular event that is taking a while?

eatyourgreens commented:
@srallen Rough values for event handlers, from an incognito window so I can rule out extensions slowing things down:
- keydown: 150ms
- blur: 50ms
- focus: 40ms

2015 MacBook Pro running Chrome 66 on OSX 10.11.6 (but why would a machine with these specs struggle with keyboard events?)

I'm baffled too because focus styling on the survey task choice questions, like 'How many animals?' works fine and those are also long lists of hidden radio buttons with styled labels.

The survey task chooser has some javascript to keep track of current focus in component state. Again, I'm not seeing any lag there when I navigate long lists via the up and down arrows. I'm not sure what's different here.

srallen commented:
@eatyourgreens It's definitely the onFocus and onBlur event handlers. I removed them and left the onChange handler for the input and the performance was much improved. I'll take a look at how the survey task is handling the focus state versus what I added in to the radio and checkbox inputs for the single and multiple choice tasks.

srallen commented:
Ok, so I went on to the master branch and added an onFocus and onBlur event and handlers to the radio button input similar to this branch and there's a similar performance hit. Each one on their own is fine most of the time, but all three of them are causing a performance issue. Doing this I've ruled out styled-components causing the performance issue.

srallen commented:
@CKrawczyk Thanks for the added information. I'll take a look at the behavior in Firefox too. I think I have a solution to help improve the performance time enough, but it'd be good to see what the cross-browser behavior is. I'll also check through the chromium issues to see if there are reports of poor performance for certain events.

srallen commented:
@eatyourgreens I've refactored the focus and blur events for the radio and checkbox inputs and they are speedy now. There are still some performance issues in Chrome however, and it appears that the event that is logged as a click (which I'm pretty sure is the onChange) can still take at least 40ms if not more. I've been able to replicate the performance issue on the master branch with just the onChange event by holding down the down arrow to go through the inputs.

I tried this out in Firefox and while I see similar performance times with the 'click'/change event, Firefox seems to be handling it more gracefully. 

I'm wondering if a debounce should be added to the onChange event? I'm not sure how to address it at this point because I'm doubting now that the changes in this branch are causing the issue.

beckyrother commented:
This looks great! Approved from me; let's merge tomorrow morning. I'll write a Talk post about it so we can post at the same time. Thanks everyone!

eatyourgreens commented:
@srallen I think JS is always going to be slow here, and the best performance would come from styling entirely in CSS, using `:focus`. That should be handled in a separate PR. This one is already so large that it’s difficult to review the whole thing. 

I’ll have another look at those event handlers today and see if I can get my head around what’s going on.

srallen commented:
@beckyrother this is requesting to merge into the classifier-light-theme branch, so it won't be deployed to production until that branch has final approval. Do you still want to announce the changes tomorrow?

srallen commented:
@eatyourgreens I agree that there is still need to follow up. I did initially just try to call `focus()` on the target label element and style using the `:focus` pseudo-class, but it wasn't working and I couldn't figure out why. But applying the data attribute manually for focus and target `[data-focus=true]` in the CSS does? 

I thought I had fixed the visual indication for focus with the selected answer with the commits yesterday. I'll double check that.

beckyrother commented:
@srallen let's wait to announce anything until the light theme is ready to go as well.

eatyourgreens commented:
Any objections to merging this?

beckyrother commented:
@eatyourgreens whenever we do it, i'll post on Talk: https://docs.google.com/document/d/1htHzK-aZw9WH4ccgvhcO5eaqzw722uteWtFr7tjIP1w/edit?usp=sharing

eatyourgreens commented:
I’m talking about merging it into #4470, so we only have one PR open.

srallen commented:
It sounds like everyone is ok with this. I'm gonna merge this and do a final staging on the classifier-light-theme branch. If there are any more issues, please comment on that PR. I still need to do a bit of manual testing, mostly in IE just to make sure there isn't anything displaying too oddly there.
